components/url_sanitizer/browser/url_sanitizer_service.cc

```diff
@@ -0,0 +1,160 @@
+    base::flat_set<std::string> params;
+    if (!CreateParamsList(params_list, &params))
+      continue;
+
+    extensions::URLPatternSet exclude_matcher;
+    CreateURLPatternSetFromValue(it.FindListPath("exclude"), &exclude_matcher);
+    auto item = std::make_unique<URLSanitizerService::MatchItem>(
+        std::move(include_matcher), std::move(exclude_matcher), params);
+
+    matchers.insert(std::move(item));
+  }
+
+  return matchers;
+}
+
+}  // namespace
+
+URLSanitizerService::URLSanitizerService() = default;
+
+URLSanitizerService::~URLSanitizerService() = default;
+
+URLSanitizerService::MatchItem::MatchItem() = default;
+URLSanitizerService::MatchItem::~MatchItem() = default;
+
+URLSanitizerService::MatchItem::MatchItem(extensions::URLPatternSet in,
+                                          extensions::URLPatternSet ex,
+                                          base::flat_set<std::string> prm) {
+  include = std::move(in);
+  exclude = std::move(ex);
+  params = std::move(prm);
+}
+
+void URLSanitizerService::Initialize(const std::string& json) {
+  base::ThreadPool::PostTaskAndReplyWithResult(
+      FROM_HERE, {base::MayBlock()}, base::BindOnce(&ParseFromJson, json),
+      base::BindOnce(&URLSanitizerService::UpdateMatchers,
+                     weak_factory_.GetWeakPtr()));
+}
+
+void URLSanitizerService::UpdateMatchers(
+    base::flat_set<std::unique_ptr<URLSanitizerService::MatchItem>> mappings) {
+  matchers_ = std::move(mappings);
+  if (initialization_callback_for_testing_)
+    std::move(initialization_callback_for_testing_).Run();
+}
+
+GURL URLSanitizerService::SanitizeURL(const GURL& initial_url) {
+  if (matchers_.empty())
+    return initial_url;
+  GURL url = initial_url;
+  for (const auto& it : matchers_) {
+    if (!it->include.MatchesURL(url) || it->exclude.MatchesURL(url))
+      continue;
+    auto sanitized_query = StripQueryParameter(url.query(), it->params);
+    GURL::Replacements replacements;
+    if (!sanitized_query.empty()) {
+      replacements.SetQueryStr(sanitized_query);
+    } else {
+      replacements.ClearQuery();
+    }
+    url = url.ReplaceComponents(replacements);
+  }
+  return url;
+}
+
+void URLSanitizerService::OnRulesReady(const std::string& content) {
+  Initialize(content);
+}
+
+// FIXME: merge with
+// browser/net/brave_site_hacks_network_delegate_helper.cc::StripQueryParameter()
+std::string URLSanitizerService::StripQueryParameter(
+    const std::string& query,
+    const base::flat_set<std::string>& trackers) {
+  const std::vector<std::string> input_kv_strings =
+      SplitString(query, "&", base::KEEP_WHITESPACE, base::SPLIT_WANT_ALL);
+  std::vector<std::string> output_kv_strings;
+  int disallowed_count = 0;
+  for (const std::string& kv_string : input_kv_strings) {
+    const std::vector<std::string> pieces = SplitString(
+        kv_string, "=", base::KEEP_WHITESPACE, base::SPLIT_WANT_NONEMPTY);
+    const std::string& key = pieces.empty() ? "" : pieces[0];
```

Comment:
Prefer using `std::string()` rather than `""`, as that results in better codegen.

Reply:
done